### PR TITLE
#12108 Speed up the twisted.web HTTP client

### LINKED ISSUE #12108: Speed up twisted.web HTTP client
The Twisted HTTP client has some code paths that aren't as optimized as one might hope, such that it should be possible to make it faster.

An initial experiment yielded at least one meaningful speedup, so I'm going to see how many more I can do.

### LINKED ISSUE #12112: Failure.__getstate__ is ignored by pickle
1. `Failure` subsclasses `BaseException`.
2. `BaseException` implements the reduce pickle API, in C.
3. As a result, you need to override `__reduce__` to impact pickling behavior.

Demonstration—notice that `Failure.__getstate__` does `c["pickled"] = 1` in its implementation in the dictionary it returns.

```python
Python 3.11.8 (main, Feb 25 2024, 16:39:33) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pickle
>>> from twisted.python.failure import Failure
>>> f = Failure(ZeroDivisionError())
>>> pickle.loads(pickle.dumps(f)).pickled
0
```

### PR DESCRIPTION
**Pull request by itamarst**
## Scope and purpose

Fixes #12108 
Fixes #12112 

Why this is faster:

* Minor reductions in general effort.
* Only calculate `Failure.parents` on demand, since it's expensive and often not used.
* `Failure.__init__` does a huge amount of unnecessary work for simple cases, so avoided it with alternative constructor. The constructor results in different pickling behavior, because of subclassing `BaseException` it seems!!!! See https://gist.github.com/itamarst/4e984bddef23a96619b37bfb929e5b24

I run my benchmark script with `hyperfine` to reduce noise from dictionary ordering randomness and the like.

Before:

```
$ hyperfine --warmup 1 "python http_client_benchmark.py"
Benchmark 1: python http_client_benchmark.py
  Time (mean ± σ):      3.992 s ±  0.056 s    [User: 3.979 s, System: 0.012 s]
  Range (min … max):    3.917 s …  4.118 s    10 runs
```

After:

```
$ hyperfine --warmup 1 "python http_client_benchmark.py"
Benchmark 1: python http_client_benchmark.py
  Time (mean ± σ):      3.266 s ±  0.032 s    [User: 3.253 s, System: 0.012 s]
  Range (min … max):    3.222 s …  3.336 s    10 runs
```

## ORIGINAL APPROACH, NOW OBSOLETE

**Important:** This relies on `Failure` objects being effectively immutable, which I _think_ is the case but should be something to consider when reviewing.

I run my benchmark script with `hyperfine` to reduce noise from dictionary ordering randomness and the like.

Before:

```
$ hyperfine --warmup 1 "python http_client_benchmark.py"
Benchmark 1: python http_client_benchmark.py
  Time (mean ± σ):      3.992 s ±  0.056 s    [User: 3.979 s, System: 0.012 s]
  Range (min … max):    3.917 s …  4.118 s    10 runs
```

After:

```
$ hyperfine --warmup 1 "python http_client_benchmark.py"
Benchmark 1: python http_client_benchmark.py
  Time (mean ± σ):      3.022 s ±  0.025 s    [User: 3.005 s, System: 0.016 s]
  Range (min … max):    2.982 s …  3.051 s    10 runs
```

Test script:

```python
from twisted.web._newclient import HTTP11ClientProtocol, Request
from twisted.internet.testing import StringTransport
from twisted.web.http_headers import Headers
import time

RESPONSE = """HTTP/1.1 200 OK
Host: blah
Foo: bar
Gaz: baz
Content-length: 3

abc""".replace("\n", "\r\n").encode("utf-8")

start = time.time()
for i in range(100_000):
    protocol = HTTP11ClientProtocol()
    protocol.makeConnection(StringTransport())
    request = Request(b"GET", b"/foo/bar", Headers({b"Host": [b"example.com"]}), None)
    d = protocol.request(request)
    protocol.dataReceived(RESPONSE)
    l = []
    d.addCallback(l.append)
    assert l
print("requests/sec:", 100_000 / (time.time() - start))
``` 

Follow-up PRs will:

- Optimize http_headers.py, which does a lot of repetitive and unnecessary work.
- Possibly split up the identity transfer encoding into two variants, one with a fixed length and one without.

## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/core/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/core/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

### CODE REVIEW

**Comment by itamarst** [comment]
please review

File: src/twisted/web/_newclient.py
Comment (adiroiban) [inline src/twisted/web/_newclient.py:932]: Can we conver this to an `if` ?

Not sure if there is a performance implicatin, but it should make the code easier to read.


```suggestion

        if self.bodyProducer.length is UNKNOWN_LENGTH:
```

File: src/twisted/web/_newclient.py
Comment (adiroiban) [inline src/twisted/web/_newclient.py:934]: Can we aslo remove this else?
Not sure if it has any performance impact. Should make the code easier to read

File: src/twisted/web/_newclient.py
Comment (adiroiban) [inline src/twisted/web/_newclient.py]: I guess that in this way we loose the traceback from the `Failure`. I think that each Failure, at initialization, will keep a copy of the  current traceback.

I guess that the main issue here is `Failure.__init__` doing all the traceback/parent search.

One suggestion.

Since the traceback is no longer usable with this refactor, would it help to have something like `twisted.python.failure.FailureWithoutTraceback` implementation, that is still a mutable failure, but has no initial traceback ?

We would still have an object initialized, but that should be faster.
I think that `Failure` without tracebacks is a valid instance.

-----------

I think that for the purpose of pefromance, it's ok not to worry about tracebacks.

But maybe we should have 2 separate  pre-created failures here.

Maybe we can have something like this, just to help with future troubleshooting and logs:

* Response body fully received in first packet.
* Response body fully received

The suggested text might not be 100% accurate. 

Since this is not really a faiulure, I think it's ok not to worry about the traceback.

File: src/twisted/web/_newclient.py
Comment (adiroiban) [inline src/twisted/web/_newclient.py]: I am not sure why we need this change.

**Review by adiroiban — COMMENTED** [review]
Many thanks for the changes.

For now, I just left this is a comment, and put back the team for review.

Happy to see performance improvements in the HTTP client.

I left a few comments inline.

As you commented in the description of the PR, my biggest concern is related to mutating the failure instance.

----------


I remember seing instances in which a failure is mutated, but I don't remember the exact code or the ticket/issue number.

I think that for the purpose of the current code, it's ok to this change.

I expect that at least some downstream application/library that relies on Twisted is mutating failures

I think I have some places in my  applications in which I mutate `Failure.value`. This was done for convenience...but I think the right thing to do is to have someting like `newFailure = oldFailure.copy(Exception('new exception'))` in which the traceback is kept.

I am doing this type of things mostly in high level application code.
The error messsage text generated by Twisted is not good enought for end users... but I still want to keep the traceback for troubleshoting 

------------

Maybe we can create a new `FrozenFailure` class and use it here.
The `FrozenFailure` will raise an error is someone tries to mutate it.

**Comment by itamarst** [comment]
It's not clear to me there _is_ a traceback in this case, since it's directly creating exceptions. Will validate tomorrow.

**Comment by adiroiban** [comment]
> It's not clear to me there is a traceback in this case, since it's directly creating exceptions. Will validate tomorrow.

You are right.
There is no traceback here, so traceback is not an issue.

Still, maybe it's a good idea to have different text messages, depending on the place from where this failure is returned.

File: src/twisted/web/_newclient.py
Comment (itamarst) [reply src/twisted/web/_newclient.py]: Because there's no `reason` variable anymore.

File: src/twisted/web/_newclient.py
Comment (itamarst) [reply src/twisted/web/_newclient.py:934]: I don't see why it'd make things more readable?

Comment (itamarst) [reply src/twisted/web/_newclient.py:934]: Oh I see, based on the previous suggestion.

File: src/twisted/web/_newclient.py
Comment (itamarst) [reply src/twisted/web/_newclient.py:932]: I tried it, I think it makes the code a little less clear.

**Comment by itamarst** [comment]
So, the main issue is mutable `Failure`. `FrozenFailure` is certainly possible, but... it'd have to make the exception frozen somehow, when it's not clear how. So a generic version is hard. And any mutable exception attributes can be mutated even if `__setattr__` is disabled... so a generic version seems difficult.

On the other hand, here it's just `ResponseDone` and `ConnectionDone` which don't have mutable attributes, really, and oughtn't be mutated. So might be doable in this case.

**Comment by itamarst** [comment]
Actually, there's a number of places that just assume immutable exceptions already, e.g. the `CONNECTION_DONE` constant, so maybe just read-only `Failure` is sufficient.

**Comment by itamarst** [comment]
OK, added a Failure subclass.

**Comment by glyph** [comment]
> OK, added a Failure subclass.

This is *probably* fine, but thinking about it does send a chill down my spine.

I don't think that an immutable failure *subclass* is actually substitutable, or thereby compliant with the letter of the compatibility policy, as it stands. I'd kinda rather make *all* Failures be immutable if that is practical, and initially enforce it at type-time, which is not covered by the compatibility policy?

**Comment by exarkun** [comment]
> > OK, added a Failure subclass.
> 
> This is _probably_ fine, but thinking about it does send a chill down my spine.
> 

*cough* CopyableFailure *cough*

> I don't think that an immutable failure subclass is actually substitutable, or thereby compliant with the letter of the compatibility policy, as it stands. I'd kinda rather make all Failures be immutable if that is practical, and initially enforce it at type-time, which is not covered by the compatibility policy?

It's certainly not.  I think _only_ enforcing it at type-time is probably good enough.  There will never be a 100% immutable Python object.  If the type checker tells someone they're not allowed to change something and they go ahead and run code that changes it anyway, that seems like a choice that's up to them.

Ideally, there'd be a generic like `typing.Readonly` - as in, `Readonly[Failure]`.  I don't think there is, but a `ReadonlyFailure` protocol that exposes all of `Failure`'s public attributes, but only for reading is still simple, just more typing.  Then I believe the definition of `freeze` is:

```
    def freeze(self) -> ReadonlyFailure: return self
```

**Comment by exarkun** [comment]
(This PR also seems like a weird place to be solving this 24 year old design mistake)

**Comment by itamarst** [comment]
https://github.com/python/typing/discussions/1123 has an example of how to define a Protocol with read-only attributes. But actually these days you can just use `Final`, which is simpler.

**Comment by itamarst** [comment]
Given the Failure object is going to end up in places, e.g. `Protocol.connectionLost` for a body protocol of a `Response`, using typing to enforce read-only-ness might end up going pretty far in terms of changing the API.

**Comment by glyph** [comment]
> _cough_ CopyableFailure _cough_

Okay, perhaps a comparable amount of chills, but it *is* a proper Liskov substitute ;)

**Comment by glyph** [comment]
> Then I believe the definition of `freeze` is:

My suggestion would be to just have `Failure` itself be frozen by default, because:

1. Mutating a failure is a pretty weird operation that I would guess rarely happens?
2. Any code doing the mutating would continue to work at runtime, unless it mutates one of these shared instances (but what mutation would such code be making, I wonder?) in which case behavior would change, but not in a technically incompatible way.
3. Any code doing the mutating would become a type error, which seems like a reasonable way for users to catch it and fix things.

If my first assumption is wrong (this would not affect much code) then we should probably do another approach. Although then I'd really want to know what the common failure-mutation patterns are.

File: src/twisted/web/_newclient.py
Comment (adiroiban) [reply src/twisted/web/_newclient.py:934]: We can leave it as it is. No problem. Thanks for looking into this.

File: src/twisted/web/newsfragments/12108.feature
Comment (adiroiban) [inline src/twisted/web/newsfragments/12108.feature]: Maybe something like this to comply with the convension for release notes

I guess that this is the main public API

```suggestion
twisted.web.client.HTTPConnectionPool used by HTTP clients now runs faster by using a little less CPU.
```

**Review by adiroiban — APPROVED** [review]
Thanks for the changes.

I only have a minor comment regarding the release notes wording.

Based on the latest discussions, I think that we can merge this PR as it is.

Also, I am also ok with merging this code with a standard `Failure`, just like it was in the first review request.

I have approved this PR, but maybe wait one day to receive more feedback from Jean-Paul and Glyph.

----------

> (This PR also seems like a weird place to be solving this 24 year old design mistake)

We don't have to make big API changes in this PR.
I think it is still usefull to discuss this issue.

I was not able to find an existing ticket/GitHub Issue discussing the read only design for Failure.

I guess we should create a new one now :)

> I'd kinda rather make all Failures be immutable if that is practical,

I think that such a change should be done in a separate PR.

So my suggestion for this PR:

* Create a new GitHub issue dedicated to the Failure read-only design (or find the existing one, if such an issue was already raised)
* Commit this PR using using `_FrozenFailure` and add a note to remove it once the Failure read-only changes are made

Thanks again!

**Comment by itamarst** [comment]
There is one existing Failure API that mutates it, https://docs.twisted.org/en/stable/api/twisted.python.failure.Failure.html#cleanFailure

**Comment by itamarst** [comment]
Based on above discussion, before proceeding on merging this PR I'm going to:

1. Open a separate issue about make `Failure` (and relevant exceptions) immutable.
2. Make a sketch on a new branch.
3. Maybe even get it to review stage, we'll see.

If it still seems plausible then we'll see what order of merging separate PRs makes sense.

**Comment by itamarst** [comment]
I opened #12111 for the immutable `Failure` case, would appreciate if people on this thread check it to see if I missed any important comments from the discussion here.

**Comment by itamarst** [comment]
I gave up on reusing `Failure` instances and went with a different approach.

**Comment by itamarst** [comment]
please review

**Comment by itamarst** [comment]
please review

**Comment by itamarst** [comment]
MyPy failures don't appear to have anything to do with this branch...

**Comment by glyph** [comment]
> MyPy failures don't appear to have anything to do with this branch...

Confirmed, fresh trunk checkout gives similar errors here. Investigating…

**Comment by glyph** [comment]
> > MyPy failures don't appear to have anything to do with this branch...
> 
> Confirmed, fresh trunk checkout gives similar errors here. Investigating…

filed https://github.com/twisted/twisted/issues/12113 to track and discuss

**Comment by glyph** [comment]
@itamarst an attempted fix is here https://github.com/twisted/twisted/pull/12115

**Comment by glyph** [comment]
OK, fix attempt seems to have worked, hopefully that can unblock this

**Comment by glyph** [comment]
Merging in the approved fix, so maybe we can get a successful run.

File: src/twisted/test/test_failure.py
Comment (glyph) [inline src/twisted/test/test_failure.py]:
```suggestion
    def test_failureWithoutTraceback(self) -> None:
```

File: src/twisted/test/test_failure.py
Comment (glyph) [inline src/twisted/test/test_failure.py]:
```suggestion
    def test_failurePickling(self) -> None:
```

File: src/twisted/test/test_failure.py
Comment (glyph) [inline src/twisted/test/test_failure.py]:
```suggestion
    def test_failurePicklingIncludesParents(self) -> None:
```

File: src/twisted/python/failure.py
Comment (glyph) [inline src/twisted/python/failure.py]:
```suggestion
    def _withoutTraceback(value: BaseException) -> Failure:
```

File: src/twisted/test/test_failure.py
Comment (glyph) [inline src/twisted/test/test_failure.py]:
```suggestion
        C{Failure._withoutTraceback(exc)} gives the same result as
```

File: src/twisted/test/test_failure.py
Comment (glyph) [inline src/twisted/test/test_failure.py]:
```suggestion
        failure2 = failure.Failure._withoutTraceback(exc)
```

File: src/twisted/test/test_failure.py
Comment (glyph) [inline src/twisted/test/test_failure.py]:
```suggestion
        C{Failure(exc)} and C{Failure._withoutTraceback(exc)} can be pickled
```

File: src/twisted/test/test_failure.py
Comment (glyph) [inline src/twisted/test/test_failure.py]:
```suggestion
        failure2 = failure.Failure._withoutTraceback(exc)
```

File: src/twisted/web/_newclient.py
Comment (glyph) [inline src/twisted/web/_newclient.py]:
```suggestion
            reason = Failure._withoutTraceback(
```

File: src/twisted/web/_newclient.py
Comment (glyph) [inline src/twisted/web/_newclient.py]:
```suggestion
            reason = Failure._withoutTraceback(
```

File: src/twisted/web/_newclient.py
Comment (glyph) [inline src/twisted/web/_newclient.py]:
```suggestion
            self._giveUp(Failure._withoutTraceback(reason))
```

**Review by glyph — COMMENTED** [review]
just some coding-standard compliance suggestions

File: src/twisted/logger/_json.py
Comment (adiroiban) [inline src/twisted/logger/_json.py:58]: Is this API used ?

In `src/twisted/logger/__init__.py` I can see only these as public API

```
from ._json import (
    eventAsJSON,
    eventFromJSON,
    jsonFileLogObserver,
    eventsFromJSONLogFile,
)

```

and I don't see any test for this code.

File: src/twisted/python/failure.py
Comment (adiroiban) [inline src/twisted/python/failure.py:422]: I know that in `trunk` we alrady have `Failure.parents` as public API and without documentation.

Can we add a few words about `Failure.parents` in this PR?

The documentation should go in `Failure.__doc__` and not here... but I can't leave a comment there.

File: src/twisted/newsfragments/12112.bugfix
Comment (adiroiban) [inline src/twisted/newsfragments/12112.bugfix]: Just a minor suggest to comply with the release notes format for bugfix

```suggestion
twisted.python.failure.Failure can now be serialized using pickling.
```

File: src/twisted/test/test_failure.py
Comment (adiroiban) [inline src/twisted/test/test_failure.py]: It looks like this code is not used. Can we removed it ?

File: src/twisted/logger/_json.py
Comment (itamarst) [reply src/twisted/logger/_json.py:58]: The tests were failing without this, so I think so.

File: src/twisted/python/failure.py
Comment (itamarst) [reply src/twisted/python/failure.py:422]: It's an implementation detail, it's better if people _don't_ use it. The fact it's public is a historic quirk...

File: src/twisted/test/test_failure.py
Comment (itamarst) [reply src/twisted/test/test_failure.py]: My worry is someone tweaks the test logic slightly and then it starts acting weirdly because of the lack of `__ne__`. But I just learned that they're optional in Python 3, lacking one it will DTRT, so will do.

File: src/twisted/python/failure.py
Comment (adiroiban) [reply src/twisted/python/failure.py:422]: Thanks for the info.

If this should be private, then maybe we need a followup issue to mark it private and raise a deprecation warning for any public usage of the API ?

**Pull request merged**
